src/_readtags.pyx

```diff
@@ -17,130 +17,142 @@ You should have received a copy of the GNU General Public License
-                return None
-            return self.c_entry.address.pattern 
-        elif key == 'lineNumber':
-            return self.c_entry.address.lineNumber 
-        elif key == 'kind':
-            if self.c_entry.kind == NULL:
-                return None
-            return self.c_entry.kind
-        elif key == 'fileScope':
-            return self.c_entry.fileScope 
-        else:
-            # It will crash if we mix NULL/0/None
-            # don't mix comparison of type
-            result = ctagsField(&self.c_entry, key)
-            if result == NULL:
-                return None
-
-            return result
+cdef create_tagEntry(const tagEntry* const c_entry):
+    cdef dict ret = {}
+    ret['name'] = c_entry.name
+    ret['file'] = c_entry.file
+    ret['fileScope'] = c_entry.fileScope
+    if c_entry.address.pattern != NULL:
+        ret['pattern'] = c_entry.address.pattern
+    if c_entry.address.lineNumber:
+        ret['lineNumber'] = c_entry.address.lineNumber
+    if c_entry.kind != NULL:
+        ret['kind'] = c_entry.kind
+    for index in range(c_entry.fields.count):
+        key = c_entry.fields.list[index].key
+        ret[key.decode()] = c_entry.fields.list[index].value
+    return ret
 
 cdef class CTags:
     cdef tagFile* file
     cdef tagFileInfo info
+    cdef tagEntry c_entry
+    cdef object current_id
 
     def __cinit__(self, filepath):
-        self.open(filepath)
+        self.file = ctagsOpen(filepath, &self.info)
+        if not self.file:
+            raise OSError(self.info.status.error_number,
+                          strerror(self.info.status.error_number),
+                          filepath)
 
     def __dealloc__(self):
-
         if self.file:
             ctagsClose(self.file)
 
     def __getitem__(self, key):
-        if key == 'opened':
-            return self.info.status.opened
-        if key == 'error_number':
-            return self.info.status.error_number
+        ret = None
         if key == 'format':
             return self.info.file.format
-        if key == 'sort':
+        elif key == 'sort':
             return self.info.file.sort
-        if key == 'author':
-            if self.info.program.author == NULL:
-                return ''
-            return self.info.program.author
-        if key == 'name':
-            if self.info.program.name == NULL:
-                return ''
-            return self.info.program.name
-        if key == 'url':
-            if self.info.program.url == NULL:
-                return ''
-            return self.info.program.url
-        if key == 'version':
-            if self.info.program.version == NULL:
-                return ''
-            return self.info.program.version
-
-
-    def open(self, filepath):
-        self.file = ctagsOpen(filepath, &self.info)
-
-        if not self.info.status.opened:
-            raise Exception('Invalid tag file')
+        else:
+            if key == 'author':
```

Comment:
I'm not sure getattr works for c structure. Maybe cython is smart enough.
I will test.